rpc, cli: add multiwallet balances rpc and use it in -getinfo

jonatack opened the pull request:
This PR implements point 2 from #17314 (show balance per wallet):

- create an RPC that returns the ismine.trusted balance for all loaded wallets
- update `bitcoin-cli -getinfo` to use the RPC

before
```json
$ bitcoin-cli -getinfo -regtest
{
  "version": 199900,
  "blocks": 15599,
  "headers": 15599,
  "verificationprogress": 1,
  "timeoffset": 0,
  "connections": 0,
  "proxy": "",
  "difficulty": 4.656542373906925e-10,
  "chain": "regtest",
  "keypoolsize": 1000,
  "paytxfee": 0.00000000,
  "balance": 0.00001000,
  "relayfee": 0.00001000
}

```
after
```json
$ bitcoin-cli -getinfo -regtest
{
  "version": 199900,
  "blocks": 15599,
  "headers": 15599,
  "verificationprogress": 1,
  "timeoffset": 0,
  "connections": 0,
  "proxy": "",
  "difficulty": 4.656542373906925e-10,
  "chain": "regtest",
  "balances": {
    "": 0.00001000,
    "Encrypted": 0.00003500,
    "day-to-day": 0.00000120,
    "side project": 0.00000094
  },
  "relayfee": 0.00001000
}
```
-----
`Review club` discussion about this PR is here: https://bitcoincore.reviews/18453.

This approach is simpler, adds no additional rpc calls in -getinfo and runs faster than doing it on the client side inside bitcoin-cli.cpp, but it also adds the first multiwallet RPC and would need to be well-considered in terms of the API.

laanwj commented:
Concept ACK

jonasschnelli commented:
Adds one more call to `getinfo`. Under the hood, it loads the same structures `GetBalance().m_mine_trusted`. Not efficient, but since performance is acceptable, I think it's worth the change to prevent from future issues and allows stripping down `getwalletinfo`

utACK f73ace05f2e67917e12026f0c69aec0e5b043366

laanwj commented:
What would make me really happy, when you're making changes here anyway, is implementing point 2 from  #17314: show balance per wallet. This is much more useful in the multi-wallet case.

promag commented:
Concept ACK.

Agree in making @laanwj happy. The change requires a `listwallets` call before the batch.

jonatack commented:
Good idea, will do.

laanwj commented:
> The change requires a listwallets call before the batch

I don't think that's necessary, `getbalances` will already retrieve the balances for all currently active wallets, which is what I mean, It doesn't need to list inactive ones in the datadir IMO.

jonatack commented:
Something like this? (getbalances version)
```json
$ bitcoin-cli -getinfo -regtest
{
  "version": 199900,
  "blocks": 15599,
  "headers": 15599,
  "verificationprogress": 1,
  "timeoffset": 0,
  "connections": 0,
  "proxy": "",
  "difficulty": 4.656542373906925e-10,
  "chain": "regtest",
  "balances": [
    {
      "": {
        "mine": {
          "trusted": 0.00001000,
          "untrusted_pending": 0.00000000,
          "immature": 0.00000000
        }
      }
    },
    {
      "encrypted": {
        "mine": {
          "trusted": 0.00003500,
          "untrusted_pending": 0.00000000,
          "immature": 0.00000000
        }
      }
    },
  ],
  "relayfee": 0.00001000
}
```
or like? (mine.trusted balances only)
```json
$ bitcoin-cli -getinfo -regtest
{
  "version": 199900,
  "blocks": 15599,
  "headers": 15599,
  "verificationprogress": 1,
  "timeoffset": 0,
  "connections": 0,
  "proxy": "",
  "difficulty": 4.656542373906925e-10,
  "chain": "regtest",
  "balances": [
    {
      "name": "",
      "amount": 0.00001000
    },
    {
      "name": "encrypted",
      "amount": 0.00003500
    }
  ],
  "relayfee": 0.00001000
}
```

laanwj commented:
It's best if `-getinfo` fits on a single screen, so I'd prefer a really condensed form, for ex.
```json
{
  "version": 199900,
  "blocks": 15599,
  "headers": 15599,
  "verificationprogress": 1,
  "timeoffset": 0,
  "connections": 0,
  "proxy": "",
  "difficulty": 4.656542373906925e-10,
  "chain": "regtest",
  "balances": {
    "": 0.00001000,
    "encrypted": 0.00003500,
  },
  "relayfee": 0.00001000
}
```

jonatack commented:
> It's best if `-getinfo` fits on a single screen, so I'd prefer a really condensed form

Perfect -- thanks :+1:

jonatack committed:
rpc: create getwalletbalances rpc, initial test coverage

jonatack committed:
cli: display multiwallet balances in -getinfo (server side version)

jonatack commented:
Done

jonatack commented:
Done. If this approach is ok, I'll add tests for the new call.

promag commented:
Why not iterate on the cli?

jonatack commented:
I could be wrong, but after after pursuing that this seemed cleaner and I think we'll ultimately want either an rpc like this, or allow passing a `multi` bool to getbalance or getbalances to do the same.

jonatack commented:
I could be wrong, but this seemed cleaner, and I think we'll ultimately want either an rpc like this or allow passing a multi bool to getbalance or getbalances to do the same.

promag commented:
Not sure about the approach TBH. I'd rather see the cli use existing calls.

The new RPC `getwalletbalances` is the first to use multiple wallets and could lead to more of multiwallet RPC. For instance, `getaddressesbylabel` could return all address of all loaded wallets.

jonatack commented:
I understand; this proposal is to get feedback.

Rebased to add tests for `getwalletbalances` in the wallet_multiwallet tests, which also improves the existing tests.

jonatack commented:
I understand; this proposal is to get feedback. Note that for now `getwalletbalances` doesn't show up in the help. Rebased to add tests for `getwalletbalances` in the wallet_multiwallet tests (which also improves the existing test coverage).

sipa commented:
I'm not opposed to having "across-all-wallets" RPCs, but perhaps it's worth making them visually distinct? How about "getallwalletbalances" or "allwalletsgetbalance"?

jonatack commented:
Yes. I was debating internally about "listbalances"...

fjahr commented:
ACK a5ea5717400d8f31ba210df5705eb6c1470da307

Reviewed code and did some light testing locally.

I think multi wallet RPCs will be useful for a lot of users. For naming I don't have strong feelings but since "multi wallet" is already the term most frequently used for this scenario verbally, I think I would prefer if that gets used if people think the distinction is necessary. So I would prefer something like "(get)multiwalletbalances" to something like listbalances or something with allwallets.

unbiasedmoose commented on src/bitcoin-cli.cpp:
```diff
@@ -239,16 +240,17 @@ class GetinfoRequestHandler: public BaseRequestHandler
         result.push_back(JSONRPCRequestObj("getnetworkinfo", NullUniValue, ID_NETWORKINFO));
         result.push_back(JSONRPCRequestObj("getblockchaininfo", NullUniValue, ID_BLOCKCHAININFO));
         result.push_back(JSONRPCRequestObj("getwalletinfo", NullUniValue, ID_WALLETINFO));
+        result.push_back(JSONRPCRequestObj("getwalletbalances", NullUniValue, ID_BALANCESINFO));
```
Super nit (please feel free to ignore): Would it make sense to use `ID_WALLETBALANCES` in place of `ID_BALANCESINFO` to preserve the `getXXX` -> `ID_XXX` naming scheme?

jonatack replied:
Thanks for the feedback. The naming of `getwalletbalances` may change but will keep this in mind.

jonatack replied:
Thanks -- done in #18574 (please review :))

unbiasedmoose commented on test/functional/wallet_multiwallet.py:
```diff
@@ -173,17 +176,34 @@ def wallet_file(name):
-        assert_equal(w1.getbalance(), 100)
-        assert_equal(w2.getbalance(), 0)
-        assert_equal(w3.getbalance(), 0)
+        wallet_name = sorted(wallet_names)[1]
```
Super nit (please feel free to ignore): Would it make sense to define a variable like `extern_wallet` next to where `wallet_names` is defined, instead of relying on the sort order of the path names?

jonatack replied:
Not a bad idea; will look at this if I have to retouch the PR.

jonatack replied:
Done in the latest rebase; it's much better with your suggestion.

Mind re-ACKing here and/or #18594 (the client-side version)?

robot-visions reviewed (commented):
ACK.  Code change looks great.

Local testing:
- Unit tests passed
- Functional tests passed (`interface_bitcoin_cli.py` and `wallet_multiwallet.py`)
- Manual smoke test passed (created multiple wallets, generated coins, made sure `bitcoin-cli -regtest -getinfo` and `bitcoin-cli -regtest getwalletbalances` behaved as expected)

Thanks for hosting the next PR Review Club!  During the session, I'd also be curious to learn when it's appropriate to add new RPCs to the `bitcoin-cli` documentation.

vasild reviewed (commented):
Approach ACK.

The modified tests `test/functional/interface_bitcoin_cli.py` and `test/functional/wallet_multiwallet.py` pass on my computer and the output of `bitcoin-cli -getinfo -regtest` looks as intended.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4264,6 +4286,7 @@ static const CRPCCommand commands[] =
     { "wallet",             "encryptwallet",                    &encryptwallet,                 {"passphrase"} },
     { "wallet",             "getaddressesbylabel",              &getaddressesbylabel,           {"label"} },
     { "wallet",             "getaddressinfo",                   &getaddressinfo,                {"address"} },
+    { "wallet",             "getwalletbalances",                &getwalletbalances,             {} },
```
If the consensus is to add a public RPC, yes, it should have documentation.

vasild replied:
Ok, makes sense. IMO the current approach where the iterate-over-wallets logic is inside bitcoind and exposed via RPC is better than having that logic inside bitcoin-cli because in the latter case it cannot be reused by RPC users - should they need that functionality they would have to replicate the logic inside their applications.

jonasschnelli commented:
Concept ACK.
I added the RPC label as this now adds a new PRC command (should probably be in the PR description).

I strongly agree with @sipa. Across-all-wallets calls should have a naming convention. IMO `allwallet[s]` should be the required string (works with `get`, `list`, `execute`, etc.). Once we have started adding across-all-wallets calls it will be difficult to change the concept (so its worth investing some brainpower in how that should work).

Another approach would be to add a special RPC endpoint like `/allwallets` that will combine the JSON output of each per-wallet-output (more a generic RPC multiwallet approach).
From the users perspective that could look like:
```
bitcoin-cli -allwallets getbalances
Output:
[
  {"name": "mywallet",
  {"output": 
		{
		  "mine": {
		    "trusted": 0.00000000,
		    "untrusted_pending": 0.00000000,
		    "immature": 0.00000000
		  }
		}
	}
,
  {"name": "encryptedwallet",
  {"output": 
		{
		  "mine": {
		    "trusted": 1.00000000,
		    "untrusted_pending": 0.00000000,
		    "immature": 0.00000000
		  }
		}
	}
,
...
]
```

I'm not convince that this is the better approach (has probably some performance and flexibility downsides), just wanted to highlight the importantness of defining the first across-all-wallets RPC.

jonatack commented:
Updated the PR name and description.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -714,6 +714,28 @@ static UniValue getreceivedbylabel(const JSONRPCRequest& request)
+        auto locked_chain = wallet.chain().lock();
+        LOCK(wallet.cs_wallet);
+
+        obj.pushKV(wallet.GetName(), ValueFromAmount(wallet.GetBalance().m_mine_trusted));
```
`GetDisplayName()` seems to be mainly for the GUI wallet. The RPC is primarily designed to be consumed by applications, whose developers would likely find `GetName()` more practical, which to me explains why GetName is used.

CLI -getinfo is between the two and oriented toward users, so the new RPC *could* return both fields, e.g. Name for RPC output and DisplayName for CLI output, but I'm ~0 on it (and admittedly also used to the RPC output as-is).

ryanofsky replied:
To be consistent with other rpcs like `listwallets` it's right for this to be using GetName not GetDisplayName. Other wallets have names, but the default wallet doesn't have a name.

theStack commented:
Concept ACK

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -714,6 +714,28 @@ static UniValue getreceivedbylabel(const JSONRPCRequest& request)
+        CWallet& wallet = *rpc_wallet;
+        wallet.BlockUntilSyncedToCurrentChain();
+
+        auto locked_chain = wallet.chain().lock();
+        LOCK(wallet.cs_wallet);
```
> A wallet's chain will never change and, I assume, all wallets' chains are the same. That is - we are operating on one chain here, for all wallets. Can we / should we lock that chain before the for-loop in order to get a consistent view (i.e. avoid getting one wallet's balance as of height H and another wallet's balance as of height H+1)?

We generally don't want wallet code to lock up the node, and after #16426 we are removing the ability of wallet code to be able to do this. It synchronization between different wallets is a concern, a solution could be to return tip hashes alongside balances. Or to restart the wallet loop if the tip changes midway.

vasild replied:
I see. So only the first question above remains (remove the locks acquisition).

ryanofsky commented:
> I strongly agree with @sipa. Across-all-wallets calls should have a naming convention. IMO `allwallet[s]` should be the required string (works with `get`, `list`, `execute`, etc.). Once we have started adding across-all-wallets calls it will be difficult to change the concept (so its worth investing some brainpower in how that should work).

Just using the word "wallets" seems pretty good as a naming convention here. And we already have a `listwallets` RPC method following this convention.
 
> Another approach would be to add a special RPC endpoint like `/allwallets` that will combine the JSON output of each per-wallet-output (more a generic RPC multiwallet approach).

This is a neat idea. Another way to implement it without a new endpoint would be with `foreachwallet` forwarding method: `bitcoin-cli foreachwallet getbalances`

jonasschnelli commented:
> Just using the word "wallets" seems pretty good as a naming convention here. And we already have a listwallets RPC method following this convention.

Could that lead to non-ideal distinguishability for certain (future?) calls (as an example `listwalletdir` versus `listwalletsdir` a.s.o.)?

> This is a neat idea. Another way to implement it without a new endpoint would be with foreachwallet forwarding method: bitcoin-cli foreachwallet getbalances

Nice idea to use a dedicated RPC method instead of an endpoint.

ryanofsky commented:
> Could that lead to non-ideal distinguishability for certain (future?) calls (as an example `listwalletdir` versus `listwalletsdir` a.s.o.)?

I think I'm not sure what the question is asking, but I'd rename `listwalletdir` to `listwalletsdir`  (and rename `-walletdir` to `-walletsdir`, see #12221), with aliases for backwards compatibility. `listwallets`, `listwalletsdir`, and `getwalletsbalances` as a convention does seems clear enough to me

jnewbery commented:
Concept ACK, but approach NACK. I don't think we should add cross-wallet RPC methods to the server. Currently we have the following convention:

For a multiwallet node:

- node RPC methods are available on the `/` endpoint
- wallet RPC methods are available on the `/wallet/<wallet_name>` endpoint
(wallet admin methods like `loadwallet`, `unloadwallet`, `listwallets` are on the node `/` endpoint

For a single wallet node:

- all node and RPC methods are available on the `/` endpoint

In future, we might want to add per-wallet permissions (#10615) or wallets in a separate process with its own RPC server (#10102). Currently, both of these things map nicely to the one-endpoint-one-wallet design we have. I think adding cross-wallet RPCs would make those things more difficult.

I think it shouldn't be too difficult to add this functionality on the bitcoin-cli client side. That seems like a better approach to me.

promag commented:
I also share @jnewbery opinion (https://github.com/bitcoin/bitcoin/pull/18453#issuecomment-605648343).

ryanofsky commented:
@jnewbery, I don't understand the distinction you're making. What makes `listwallets` and `unloadwallet` "wallet admin" methods that are ok to have, and `getwalletsbalances` not a wallet admin method, or not a wallet admin method that's ok to have?

I also don't see this interfering in a significant way with #10102 or #10615. From perspective of #10102 at least, this isn't different than the current `listwallets` method.

jnewbery commented:
>  I don't understand the distinction you're making. What makes listwallets and unloadwallet "wallet admin" methods that are ok to have

I think it's fine to allow the node administrator to see which wallets are loaded or to unload a wallet, but they shouldn't be able to carry out any actions on that wallet. It would be unexpected to add wallet permissions to a wallet and then find out that a user without those permissions is able to see the balance. How would you expect this (or other cross-wallet RPC methods) to behave if some of the wallets had permissions?

I generally don't like arguments that we shouldn't add a new feature because it might interact badly with some theoretical future feature, but in this case I don't see the need to add a whole new class of cross-wallet RPCs when the functionality can easily be added to the client.

jonasschnelli commented:
> Concept ACK, but approach NACK. I don't think we should add cross-wallet RPC methods to the server.

Thanks @jnewbery for sharing your thoughts.
The advantage of having serverside across-all-wallets calls if probably only efficiency (performance, bandwidth) which might be not worth the effort. I think even with per wallet permissions serverside  across-all-wallet calls would work as well (some elements may be omitted or a permission error is returned).

**Our JSON RPC implementation supports batches. Though,... I guess since we are using different endpoints for the wallet selection, batching across wallets is not possible?**

Therefore I think it would make sense to support some sort of batching option that allow across-wallets batches (for performance reasons).

@ryanofsky `foreachwallet` idea seems to be a viable way. Maybe as an endpoint or a method that has sub-methods are parameter.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -239,16 +240,17 @@ class GetinfoRequestHandler: public BaseRequestHandler
-        std::vector<UniValue> batch = JSONRPCProcessBatchReply(batch_in, 3);
-        // Errors in getnetworkinfo() and getblockchaininfo() are fatal, pass them on
-        // getwalletinfo() is allowed to fail in case there is no wallet.
+        std::vector<UniValue> batch = JSONRPCProcessBatchReply(batch_in, 4);
```
Thanks -- done in #18574 (please ack :).

vasild replied:
Good idea to split the PR in two.

jonatack commented:
I've extracted the original PR to #18574 to preserve this one as a discussion on multiwallet RPC/CLI.

jonatack commented:
After looking at #10102 (and #10615 which is closed), I tend to agree with @ryanofsky and @jonasschnelli. Will propose multiwallet balances for -getinfo inside bitcoin-cli.cpp as a lower-scope solution for now in a separate PR, while this discussion continues in parallel for multiwallets on the server side.

jonatack committed:
test: add multiwallet -getinfo tests to interface_bitcoin_cli.py

jonatack commented:
Rebased.

I implemented a client-side version of this PR at #18594 per the approach suggested by @laanwj, @promag, and @jnewbery. For now I think it's the immediate solution, and once a server-side multiwallet RPC API is adopted, it can be used for -getinfo.

jonatack commented:
Rebased, with improved tests.

jb55 commented:
@jonasschnelli:
> Therefore I think it would make sense to support some sort of batching option that allow across-wallets batches (for performance reasons).

ACK this. Right now I have bash scripts for doing multi-wallet things like balances and coins but it's pretty horrible and slow since I can't batch requests. It would be so much nicer if there was a special wallet endpoint you could hit which signals that you want the result from all wallets, or at least a way to batch requests for different wallets.

jonatack commented:
Rebased on master for the CI fix.

luke-jr reviewed (changes requested):
Concept NACK specifically to enabling access to the wallet list (or non-selected wallets) for existing interfaces.

Up until now, RPCs would only ever access the wallet the request was sent to...

jonatack commented:
Thanks to everyone for reviewing.

Summary:
4 Concept ACKs
1 Approach ACK
1 Concept NACK
2 Approach NACKs
4 ACKs

To take the feedback into account, a client-side, smaller-scope version of this feature for `-getinfo` only, with Concept ACKs, ACKs, and several rounds of review, is ready now for final review at #18594. Can you please review?
